DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,304 @@
+
+Herein lies an exception. When code is aware it is executing under a `main` func, or a derivative of `main` such as a `Command` object, it may have grounds for panicking if it identifies the program is invalid. Program invalidity is typically representative of a programmer error. For example, consider a `Command` object that requires, under certain circumstances, a `foo` flag to be configured as part of CLI argument parsing. If the program finds that flag doesn't exist (not to be confused with "wasn't set by the user"), the program is invalid. This is indicative of a programmer error as they _forgot_ to add the CLI flag. In this instance it is impossible for a user to fix the problem, because the program is invalid. Consequently, the program has grounds to panic. It is important to recognize that the panicking code is _fully aware_ of its execution context.
+
+## Style
+
+### Variable declaration
+
+When declaring and not initializing, prefer `var`. For example, `var vehicle Vehicle`.
+
+When delcaring and initializing, use `:=`. For example, `vehicle := NewVehicle()`
+
+Make intent clear for complicated variable initializations (contradictory to the 2 above rules). For example `var p uint32 = 0x80` instead of `p := uint32(0x80)`.
+
+### Line length
+
+Traditionally, many coding standards have stipulated 80 chars as the maximum line length. It is thought to originate from the days of punch cards where IBM used 80 column widths. The width translated well to small width terminal monitors hence was adopted in the early days of computing.
+
+In the present day we still see 80 chars line length feature in coding standards. However, much has changed since the standard was originally employed. Attempts, through research, to pin down the optimal line length exists but have resulted in conjecture and contradiction. For this reason, we consider other use-cases and constraints to help decide line length:
+
+- laptops with less than 16" wide monitors are in abundance.
+- side-by-side comparison of files is useful.
+- horrizontal scrolling is annoying.
+- some developers _need_ larger font sizes.
+
+When developing code, we ask you to take these points into consideration and not to create obnoxiously long lines. Compliment exisiting code. Configure your IDE to plot margins so it may aid your decision. A good rule of thumb is the 100 characters mark give or take 20.
+
+### File naming
+
+All `go` file names must follow `snake_case`. Using snake case compliments the Go compiler that expects test files to be appended with `_test`.
+
+## References
+
+These references are the basis for this document.
+
+- Practical Go - Dave Cheney - https://dave.cheney.net/practical-go/presentations/gophercon-israel.html
+- Best Practices - Peter Bourgon - https://peter.bourgon.org/go-best-practices-2016/
+- Effective Go - The Go Team - https://go.dev/doc/effective_go
+- Code Review Comments - The Go Community - https://github.com/golang/go/wiki/CodeReviewComments
+
+## To do
```

Comment:
would love to see more about logging -- when, what, writing concise/clear log messages; really important for building for operations, especially since we'll be debugging a lot of this stuff 'blind' without access to the cluster in most cases

Reply:
Yea great point. I've written a 1 pager on logging previously because of how important it is. I'll add another section in the to do.

Reply:
Added the todo.